### ISSUE TITLE
No MSD on Windows 7

### ISSUE THREAD

**Issue by rusefillc** [rusefi/rusefi#2536]
I have rusEFI SD card MSD working on Windows 10 but not on Windows 7

**Comment by mck1117** [comment]
Can you snip the relevant section device manager with it set to "Devices by connection"? I see this in w10:

![image](https://user-images.githubusercontent.com/568254/114097185-8b8f8880-9874-11eb-853f-f9ee79606dda.png)

**Comment by rusefillc** [comment]
not what you want

![image](https://user-images.githubusercontent.com/48498823/114097279-d2d64300-988d-11eb-9b0e-607a9bf02685.png)

**Comment by rusefillc** [comment]
I do not see it

![image](https://user-images.githubusercontent.com/48498823/114097472-129d2a80-988e-11eb-9544-d78ff51e591c.png)

**Comment by rusefillc** [comment]
Correction

![image](https://user-images.githubusercontent.com/48498823/114097538-2ba5db80-988e-11eb-8c6f-45d09e036521.png)

**Comment by mck1117** [comment]
Download this, run it, and dump the descriptor data from the ECU device.

https://www.uwe-sieber.de/files/UsbTreeView_x64.zip

![image](https://user-images.githubusercontent.com/568254/114097814-6b13fe00-9875-11eb-87a9-2800a1b6bd8f.png)

Throw the contents of the right side up on gist.github.com or something.

**Comment by rusefillc** [comment]
![image](https://user-images.githubusercontent.com/48498823/114098427-52184680-988f-11eb-8b8c-d5758c3f4de2.png)

```

    =========================== USB Port3 ===========================

Connection Status        : 0x01 (Device is connected)
Port Chain               : 1-1-3

      ========================== Summary =========================
Vendor ID                : 0x0483 (STMicroelectronics)
Product ID               : 0x5740
USB version              : 1.10
Port maximum Speed       : High-Speed
Device maximum Speed     : Full-Speed
Device Connection Speed  : Full-Speed
Self Powered             : yes
Demanded Current         : 200 mA
Used Endpoints           : 6

      ======================== USB Device ========================

        +++++++++++++++++ Device Information ++++++++++++++++++
Friendly Name            : STMicroelectronics Virtual COM Port (COM30)
Device Description       : STMicroelectronics Virtual COM Port
Device Path              : \\?\USB#VID_0483&PID_5740#3F0029000A50504750323520#{a5dcbf10-6530-11d2-901f-00c04fb951ed} (GUID_DEVINTERFACE_USB_DEVICE)
Kernel Name              : \Device\USBPDO-6
Device ID                : USB\VID_0483&PID_5740\3F0029000A50504750323520
Hardware IDs             : USB\VID_0483&PID_5740&REV_0200 USB\VID_0483&PID_5740
Driver KeyName           : {4d36e978-e325-11ce-bfc1-08002be10318}\0028 (GUID_DEVCLASS_PORTS)
Driver                   : system32\DRIVERS\usbser.sys (Version: 6.1.7601.18247  Date: 2013-08-28)
Driver Inf               : C:\Windows\inf\oem23.inf
Legacy BusType           : PNPBus
Class                    : Ports
Class GUID               : {4d36e978-e325-11ce-bfc1-08002be10318} (GUID_DEVCLASS_PORTS)
Service                  : usbser
Enumerator               : USB
Location Info            : Port_#0003.Hub_#0004
Location IDs             : PCIROOT(0)#PCI(1D00)#USBROOT(0)#USB(1)#USB(3)
Container ID             : {a21f8114-4c23-52fc-89af-85ebb3abfc20}
Manufacturer Info        : STMicroelectronics.
Capabilities             : 0x94 (Removable, UniqueID, SurpriseRemovalOK)
Status                   : 0x0180600A (DN_DRIVER_LOADED, DN_STARTED, DN_DISABLEABLE, DN_REMOVABLE, DN_NT_ENUMERATOR, DN_NT_DRIVER)
Problem Code             : 0
HcDisableSelectiveSuspend: 0
EnableSelectiveSuspend   : 0
SelectiveSuspendEnabled  : 0
EnhancedPowerMgmtEnabled : 0
IdleInWorkingState       : 0
WakeFromSleepState       : 0
Power State              : D0 (supported: D0, D3, wake from D0)
COM-Port                 : COM30 (\Device\USBSER000)

        ---------------- Connection Information ---------------
Connection Index         : 0x03 (Port 3)
Connection Status        : 0x01 (DeviceConnected)
Current Config Value     : 0x01 (Configuration 1)
Device Address           : 0x03 (3)
Is Hub                   : 0x00 (no)
Device Bus Speed         : 0x01 (Full-Speed)
Number Of Open Pipes     : 0x05 (5 pipes to data endpoints)
Pipe[0]                  : EndpointID=1  Direction=IN   ScheduleOffset=0  Type=Bulk
Pipe[1]                  : EndpointID=1  Direction=OUT  ScheduleOffset=0  Type=Bulk
Pipe[2]                  : EndpointID=3  Direction=IN   ScheduleOffset=3  Type=Interrupt
Pipe[3]                  : EndpointID=2  Direction=OUT  ScheduleOffset=0  Type=Bulk
Pipe[4]                  : EndpointID=2  Direction=IN   ScheduleOffset=0  Type=Bulk
Data (HexDump)           : 03 00 00 00 12 01 10 01 EF 02 01 40 83 04 40 57   ...........@..@W
                           00 02 01 02 03 01 01 01 00 03 00 05 00 00 00 01   ................
                           00 00 00 07 05 81 02 40 00 00 00 00 00 00 07 05   .......@........
                           01 02 40 00 00 00 00 00 00 07 05 83 03 08 00 FF   ..@.............
                           03 00 00 00 07 05 02 02 40 00 00 00 00 00 00 07   ........@.......
                           05 82 02 40 00 00 00 00 00 00                     ...@......

    ---------------------- Device Descriptor ----------------------
bLength                  : 0x12 (18 bytes)
bDescriptorType          : 0x01 (Device Descriptor)
bcdUSB                   : 0x110 (USB Version 1.10)
bDeviceClass             : 0xEF (Miscellaneous)
bDeviceSubClass          : 0x02
bDeviceProtocol          : 0x01 (IAD - Interface Association Descriptor)
bMaxPacketSize0          : 0x40 (64 bytes)
idVendor                 : 0x0483 (STMicroelectronics)
idProduct                : 0x5740
bcdDevice                : 0x0200
iManufacturer            : 0x01 (String Descriptor 1)
 Language 0x0409         : "rusEFI LLC"
iProduct                 : 0x02 (String Descriptor 2)
 Language 0x0409         : "rusEFI Engine Management ECU"
iSerialNumber            : 0x03 (String Descriptor 3)
 Language 0x0409         : "3F0029000A50504750323520"
bNumConfigurations       : 0x01 (1 Configuration)
Data (HexDump)           : 12 01 10 01 EF 02 01 40 83 04 40 57 00 02 01 02   .......@..@W....
                           03 01                                             ..

    ------------------ Configuration Descriptor -------------------
bLength                  : 0x09 (9 bytes)
bDescriptorType          : 0x02 (Configuration Descriptor)
wTotalLength             : 0x0062 (98 bytes)
bNumInterfaces           : 0x03 (3 Interfaces)
bConfigurationValue      : 0x01 (Configuration 1)
iConfiguration           : 0x00 (No String Descriptor)
bmAttributes             : 0xC0
 D7: Reserved, set 1     : 0x01
 D6: Self Powered        : 0x01 (yes)
 D5: Remote Wakeup       : 0x00 (no)
 D4..0: Reserved, set 0  : 0x00
MaxPower                 : 0x64 (200 mA)
Data (HexDump)           : 09 02 62 00 03 01 00 C0 64 09 04 00 00 02 08 06   ..b.....d.......
                           50 00 07 05 81 02 40 00 00 07 05 01 02 40 00 00   P.....@......@..
                           08 0B 01 02 02 02 01 02 09 04 01 00 01 02 02 01   ................
                           00 05 24 00 10 01 05 24 01 00 02 04 24 02 02 05   ..$....$....$...
                           24 06 01 02 07 05 83 03 08 00 FF 09 04 02 00 02   $...............
                           0A 00 00 00 07 05 02 02 40 00 00 07 05 82 02 40   ........@......@
                           00 00                                             ..

        ---------------- Interface Descriptor -----------------
bLength                  : 0x09 (9 bytes)
bDescriptorType          : 0x04 (Interface Descriptor)
bInterfaceNumber         : 0x00
bAlternateSetting        : 0x00
bNumEndpoints            : 0x02 (2 Endpoints)
bInterfaceClass          : 0x08 (Mass Storage)
bInterfaceSubClass       : 0x06 (SCSI transparent command set)
bInterfaceProtocol       : 0x50 (Bulk-Only Transport)
iInterface               : 0x00 (No String Descriptor)
Data (HexDump)           : 09 04 00 00 02 08 06 50 00                        .......P.

        ----------------- Endpoint Descriptor -----------------
bLength                  : 0x07 (7 bytes)
bDescriptorType          : 0x05 (Endpoint Descriptor)
bEndpointAddress         : 0x81 (Direction=IN EndpointID=1)
bmAttributes             : 0x02 (TransferType=Bulk)
wMaxPacketSize           : 0x0040 (64 bytes)
bInterval                : 0x00 (ignored)
Data (HexDump)           : 07 05 81 02 40 00 00                              ....@..

        ----------------- Endpoint Descriptor -----------------
bLength                  : 0x07 (7 bytes)
bDescriptorType          : 0x05 (Endpoint Descriptor)
bEndpointAddress         : 0x01 (Direction=OUT EndpointID=1)
bmAttributes             : 0x02 (TransferType=Bulk)
wMaxPacketSize           : 0x0040 (64 bytes)
bInterval                : 0x00 (ignored)
Data (HexDump)           : 07 05 01 02 40 00 00                              ....@..

        ------------------- IAD Descriptor --------------------
bLength                  : 0x08 (8 bytes)
bDescriptorType          : 0x0B
bFirstInterface          : 0x01
bInterfaceCount          : 0x02
bFunctionClass           : 0x02 (Communications and CDC Control)
bFunctionSubClass        : 0x02
bFunctionProtocol        : 0x01
iFunction                : 0x02 (String Descriptor 2)
 Language 0x0409         : "rusEFI Engine Management ECU"
Data (HexDump)           : 08 0B 01 02 02 02 01 02                           ........

        ---------------- Interface Descriptor -----------------
bLength                  : 0x09 (9 bytes)
bDescriptorType          : 0x04 (Interface Descriptor)
bInterfaceNumber         : 0x01
bAlternateSetting        : 0x00
bNumEndpoints            : 0x01 (1 Endpoint)
bInterfaceClass          : 0x02 (Communications and CDC Control)
bInterfaceSubClass       : 0x02 (Abstract Control Model)
bInterfaceProtocol       : 0x01 (AT Commands defined by ITU-T V.250 etc)
iInterface               : 0x00 (No String Descriptor)
Data (HexDump)           : 09 04 01 00 01 02 02 01 00                        .........

        -------------- CDC Interface Descriptor ---------------
bFunctionLength          : 0x05 (5 bytes)
bDescriptorType          : 0x24 (Interface)
bDescriptorSubType       : 0x00 (Header Functional Descriptor)
bcdCDC                   : 0x110 (CDC Version 1.10)
Data (HexDump)           : 05 24 00 10 01                                    .$...

        -------------- CDC Interface Descriptor ---------------
bFunctionLength          : 0x05 (5 bytes)
bDescriptorType          : 0x24 (Interface)
bDescriptorSubType       : 0x01 (Call Management Functional Descriptor)
bmCapabilities           : 0x00
 D7..2                   : 0x00 (Reserved)
 D1                      : 0x00 (sends/receives call management information only over the Communication Class interface)
 D0                      : 0x00 (does not handle call management itself)
bDataInterface           : 0x02
Data (HexDump)           : 05 24 01 00 02                                    .$...

        -------------- CDC Interface Descriptor ---------------
bFunctionLength          : 0x04 (4 bytes)
bDescriptorType          : 0x24 (Interface)
bDescriptorSubType       : 0x02 (Abstract Control Management Functional Descriptor)
bmCapabilities           : 0x02
 D7..4                   : 0x00 (Reserved)
 D3                      : 0x00 (not supports the notification Network_Connection)
 D2                      : 0x00 (not supports the request Send_Break)
 D1                      : 0x01 (supports the request combination of Set_Line_Coding, Set_Control_Line_State, Get_Line_Coding, and the notification Serial_State)
 D0                      : 0x00 (not supports the request combination of Set_Comm_Feature, Clear_Comm_Feature, and Get_Comm_Feature)
Data (HexDump)           : 04 24 02 02                                       .$..

        -------------- CDC Interface Descriptor ---------------
bFunctionLength          : 0x05 (5 bytes)
bDescriptorType          : 0x24 (Interface)
bDescriptorSubType       : 0x06 (Union Functional Descriptor)
bControlInterface        : 0x01
bSubordinateInterface[0] : 0x02
Data (HexDump)           : 05 24 06 01 02                                    .$...

        ----------------- Endpoint Descriptor -----------------
bLength                  : 0x07 (7 bytes)
bDescriptorType          : 0x05 (Endpoint Descriptor)
bEndpointAddress         : 0x83 (Direction=IN EndpointID=3)
bmAttributes             : 0x03 (TransferType=Interrupt)
wMaxPacketSize           : 0x0008 (8 bytes)
bInterval                : 0xFF (255 ms)
Data (HexDump)           : 07 05 83 03 08 00 FF                              .......

        ---------------- Interface Descriptor -----------------
bLength                  : 0x09 (9 bytes)
bDescriptorType          : 0x04 (Interface Descriptor)
bInterfaceNumber         : 0x02
bAlternateSetting        : 0x00
bNumEndpoints            : 0x02 (2 Endpoints)
bInterfaceClass          : 0x0A (CDC-Data)
bInterfaceSubClass       : 0x00
bInterfaceProtocol       : 0x00
iInterface               : 0x00 (No String Descriptor)
Data (HexDump)           : 09 04 02 00 02 0A 00 00 00                        .........

        ----------------- Endpoint Descriptor -----------------
bLength                  : 0x07 (7 bytes)
bDescriptorType          : 0x05 (Endpoint Descriptor)
bEndpointAddress         : 0x02 (Direction=OUT EndpointID=2)
bmAttributes             : 0x02 (TransferType=Bulk)
wMaxPacketSize           : 0x0040 (64 bytes)
bInterval                : 0x00 (ignored)
Data (HexDump)           : 07 05 02 02 40 00 00                              ....@..

        ----------------- Endpoint Descriptor -----------------
bLength                  : 0x07 (7 bytes)
bDescriptorType          : 0x05 (Endpoint Descriptor)
bEndpointAddress         : 0x82 (Direction=IN EndpointID=2)
bmAttributes             : 0x02 (TransferType=Bulk)
wMaxPacketSize           : 0x0040 (64 bytes)
bInterval                : 0x00 (ignored)
Data (HexDump)           : 07 05 82 02 40 00 00                              ....@..

      -------------------- String Descriptors -------------------
             ------ String Descriptor 0 ------
bLength                  : 0x04 (4 bytes)
bDescriptorType          : 0x03 (String Descriptor)
Language ID[0]           : 0x0409 (English - United States)
Data (HexDump)           : 04 03 09 04                                       ....
             ------ String Descriptor 1 ------
bLength                  : 0x16 (22 bytes)
bDescriptorType          : 0x03 (String Descriptor)
Language 0x0409          : "rusEFI LLC"
Data (HexDump)           : 16 03 72 00 75 00 73 00 45 00 46 00 49 00 20 00   ..r.u.s.E.F.I. .
                           4C 00 4C 00 43 00                                 L.L.C.
             ------ String Descriptor 2 ------
bLength                  : 0x3A (58 bytes)
bDescriptorType          : 0x03 (String Descriptor)
Language 0x0409          : "rusEFI Engine Management ECU"
Data (HexDump)           : 3A 03 72 00 75 00 73 00 45 00 46 00 49 00 20 00   :.r.u.s.E.F.I. .
                           45 00 6E 00 67 00 69 00 6E 00 65 00 20 00 4D 00   E.n.g.i.n.e. .M.
                           61 00 6E 00 61 00 67 00 65 00 6D 00 65 00 6E 00   a.n.a.g.e.m.e.n.
                           74 00 20 00 45 00 43 00 55 00                     t. .E.C.U.
             ------ String Descriptor 3 ------
bLength                  : 0x32 (50 bytes)
bDescriptorType          : 0x03 (String Descriptor)
Language 0x0409          : "3F0029000A50504750323520"
Data (HexDump)           : 32 03 33 00 46 00 30 00 30 00 32 00 39 00 30 00   2.3.F.0.0.2.9.0.
                           30 00 30 00 41 00 35 00 30 00 35 00 30 00 34 00   0.0.A.5.0.5.0.4.
                           37 00 35 00 30 00 33 00 32 00 33 00 35 00 32 00   7.5.0.3.2.3.5.2.
                           30 00                                             0.

```

**Comment by mck1117** [comment]
Tl;dr, it's impossible on Win7, and win7 is officially out of support by Microsoft.
